File: 813-all-paths-from-source-to-target/solution.py
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        
        def dfs(src, path):

            path.append(src)

            if src == len(graph) - 1:
                res.append(path[:])
                path.pop(-1)
                return
            
            for node in graph[src]:
                dfs(node, path)
        
            path.pop(-1)

        res = []
        dfs(0, list())
        return res

File: 813-all-paths-from-source-to-target/solution.cpp
class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        int target = int(graph.size()) - 1;
        vector<int> path;
        vector<vector<int>> results;

        function<void(int, vector<int>&)> backtrack = [&](int currNode, vector<int>& path) {
            path.push_back(currNode);
            if (currNode == target) {
                results.push_back(vector(path));
                path.pop_back();
                return;
            }
            for (int nextNode : graph[currNode]) {
                backtrack(nextNode, path);
            }
            path.pop_back();
        };

        backtrack(0, path);
        return results;
    }
};
